pkg/cmd/release-payload-controller/release_creation_status_controller.go

```diff
@@ -0,0 +1,315 @@
+
+	klog.Info("Starting Release Creation Status Controller")
+	defer func() {
+		klog.Info("Shutting down Release Creation Status Controller")
+		c.queue.ShutDown()
+		klog.Info("Release Creation Status Controller shut down")
+	}()
+
+	if !cache.WaitForNamedCacheSync("ReleaseCreationStatusController", ctx.Done(), c.cachesToSync...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *ReleaseCreationStatusController) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *ReleaseCreationStatusController) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(ctx, key.(string))
+
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %w", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *ReleaseCreationStatusController) sync(ctx context.Context, key string) error {
+	klog.V(4).Infof("Starting ReleaseCreationStatusController sync")
+	defer klog.V(4).Infof("ReleaseCreationStatusController sync done")
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	klog.V(4).Infof("Processing ReleasePayload: '%s/%s' from workQueue", namespace, name)
+
+	// Get the ReleasePayload resource with this namespace/name
+	originalReleasePayload, err := c.releasePayloadLister.ReleasePayloads(namespace).Get(name)
+	// The ReleasePayload resource may no longer exist, in which case we stop processing.
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// If the release creation job status is terminal (Success), then we have noting else to do
+	if len(originalReleasePayload.Status.ReleaseCreationJobResult.Status) > 0 && originalReleasePayload.Status.ReleaseCreationJobResult.Status == v1alpha1.ReleaseCreationJobSuccess {
+		return nil
+	}
+
+	if len(originalReleasePayload.Status.ReleaseCreationJobResult.Coordinates.Namespace) == 0 || len(originalReleasePayload.Status.ReleaseCreationJobResult.Coordinates.Name) == 0 {
+		return ErrCoordinatesNotSet
+	}
+
+	// Lookup the job. If not found, then the status should be unknown...
+	// TODO: consider a timeout...
+	job, err := c.batchJobLister.Jobs(originalReleasePayload.Status.ReleaseCreationJobResult.Coordinates.Namespace).Get(originalReleasePayload.Status.ReleaseCreationJobResult.Coordinates.Name)
+	if k8serrors.IsNotFound(err) {
+		klog.V(4).Infof("Unable to locate release creation job: %s/%s", c.batchJobNamespace, originalReleasePayload.Name)
+		// Reset the error to allow for further processing
+		err = nil
+		// Instantiating a nil Job to process, to force the logic to set status to "Unknown"
```

Comment:
this isn't a nil job.

You're creating a zero-value Job so that you can have a single flow of logic based on finding completion markers in job status that will not be present in a zero value job.

FWIW, in less structured and reviewed APIs, this is risky and you should instead write the values you mean into status.

Reply:
Converted logic to use a boolean flag and directly setting Status and Message.